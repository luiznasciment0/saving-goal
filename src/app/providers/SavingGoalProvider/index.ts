import constate from 'constate'
import { useReducer, useCallback } from 'react'

import { MinDate, SavingGoalState, Actions } from 'app/models/SavingGoal'

const currencyMask = (money: number) => {
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2
  }).format(money)
}

const reducer = (state: SavingGoalState, action: Actions) => {
  switch (action.type) {
    case 'SET_TOTAL_AMOUNT':
      return { ...state, totalAmount: action.money }
    case 'SET_DATE':
      return { ...state, reachDate: action.date }
    case 'SET_MONTHLY_AMOUNT':
      return {
        ...state,
        monthlyAmount: action.amount
      }
  }
}

const useSavingGoal = ({ initialDate }: MinDate) => {
  const [state, dispatch] = useReducer(reducer, {
    reachDate: initialDate,
    totalAmount: '',
    monthlyAmount: ''
  })

  const setAmount = (money: string) => {
    dispatch({ type: 'SET_TOTAL_AMOUNT', money })
  }

  const setDateToReachGoal = (date: Date) => {
    dispatch({ type: 'SET_DATE', date })
  }

  const calcMonthlyAmount = useCallback(
    (dateDiff: number, totalAmount: string) => {
      const initialAmount = totalAmount === '0' ? '$0,00' : `${totalAmount}`

      if (dateDiff === 0) {
        dispatch({ type: 'SET_MONTHLY_AMOUNT', amount: initialAmount })
        return
      }

      const numberTotalAmount = Number(totalAmount.replace(/[,]+/g, ''))

      dispatch({
        type: 'SET_MONTHLY_AMOUNT',
        amount: currencyMask(numberTotalAmount / dateDiff)
      })
    },
    []
  )

  return {
    state,
    effects: {
      setAmount,
      setDateToReachGoal,
      calcMonthlyAmount
    }
  }
}

const [SavingGoalProvider, useStateSavingGoal, useEffectsSavingGoal] = constate(
  useSavingGoal,
  (value) => value.state,
  (value) => value.effects
)

export { SavingGoalProvider, useStateSavingGoal, useEffectsSavingGoal }
