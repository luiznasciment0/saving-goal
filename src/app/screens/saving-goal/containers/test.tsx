import { render, fireEvent } from '@testing-library/react'
import 'jest-styled-components'

import { SavingGoalProvider } from 'app/providers/SavingGoalProvider'
import SavingGoalContainer from '.'

const initialDate = new Date(2021, 1)

const setup = () => {
  const utils = render(
    <SavingGoalProvider initialDate={initialDate}>
      <SavingGoalContainer initialDate={initialDate} />
    </SavingGoalProvider>
  )

  const moneyInput = utils.getByLabelText('Money input')

  const datePicker = utils.getByLabelText('Month and year')
  const buttonRight = utils.getByLabelText('Next month')
  const buttonLeft = utils.getByLabelText('Previous month')

  const savingPlanTitle = utils.getByLabelText(
    'Amount of money to be monthly saved'
  )

  const savingPlanDetails = utils.getByLabelText('Monthly save details')

  return {
    moneyInput,
    datePicker,
    buttonLeft,
    buttonRight,
    savingPlanTitle,
    savingPlanDetails,
    ...utils
  }
}

describe('<SavingGoalContainer />', () => {
  it('should work correctly', () => {
    const {
      container,
      moneyInput,
      datePicker,
      savingPlanTitle,
      savingPlanDetails,
      buttonLeft,
      buttonRight
    } = setup()

    expect(moneyInput).toHaveValue('')
    expect(datePicker).toHaveTextContent('February2021')
    expect(savingPlanTitle).toHaveTextContent('Monthly amount$0.00')
    expect(savingPlanDetails).toHaveTextContent(
      'You’re planning 1 monthly deposits to reach your $0.00 goal by February 2021.'
    )

    fireEvent.change(moneyInput, { target: { value: '100000' } })
    expect(moneyInput).toHaveValue('1,000.00')
    expect(savingPlanTitle).toHaveTextContent('Monthly amount$1,000.00')
    expect(savingPlanDetails).toHaveTextContent(
      'You’re planning 1 monthly deposits to reach your $1,000.00 goal by February 2021.'
    )

    fireEvent.keyDown(buttonRight, { key: 'ArrowRight', code: 'ArrowRight' })
    expect(datePicker).toHaveTextContent('March2021')
    expect(savingPlanTitle).toHaveTextContent('Monthly amount$500.00')
    expect(savingPlanDetails).toHaveTextContent(
      'You’re planning 2 monthly deposits to reach your $1,000.00 goal by March 2021.'
    )

    fireEvent.keyDown(buttonRight, { key: 'ArrowRight', code: 'ArrowRight' })
    expect(datePicker).toHaveTextContent('April2021')
    expect(savingPlanTitle).toHaveTextContent('Monthly amount$333.33')
    expect(savingPlanDetails).toHaveTextContent(
      'You’re planning 3 monthly deposits to reach your $1,000.00 goal by April 2021.'
    )

    fireEvent.keyDown(buttonLeft, { key: 'ArrowLeft', code: 'ArrowLeft' })
    expect(datePicker).toHaveTextContent('March2021')
    expect(savingPlanTitle).toHaveTextContent('Monthly amount$500.00')
    expect(savingPlanDetails).toHaveTextContent(
      'You’re planning 2 monthly deposits to reach your $1,000.00 goal by March 2021.'
    )

    fireEvent.click(buttonRight)
    expect(datePicker).toHaveTextContent('April2021')
    expect(savingPlanTitle).toHaveTextContent('Monthly amount$333.33')
    expect(savingPlanDetails).toHaveTextContent(
      'You’re planning 3 monthly deposits to reach your $1,000.00 goal by April 2021.'
    )

    fireEvent.click(buttonLeft)
    expect(datePicker).toHaveTextContent('March2021')
    expect(savingPlanTitle).toHaveTextContent('Monthly amount$500.00')
    expect(savingPlanDetails).toHaveTextContent(
      'You’re planning 2 monthly deposits to reach your $1,000.00 goal by March 2021.'
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
