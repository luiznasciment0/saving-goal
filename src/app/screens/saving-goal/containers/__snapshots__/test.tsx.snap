// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SavingGoalContainer /> should work correctly 1`] = `
.c1 {
  font-size: 1.2rem;
  line-height: 1.6rem;
  font-family: 'Work Sans';
  color: #1e2a32;
  font-weight: 400;
}

.c8 {
  font-size: 1.4rem;
  line-height: 2.1rem;
  font-family: 'Work Sans';
  color: #1e2a32;
  font-weight: 600;
}

.c9 {
  font-size: 1.4rem;
  line-height: 2.1rem;
  font-family: 'Work Sans';
  color: #708797;
  font-weight: 400;
}

.c5 {
  border: 1px solid #e9eef2;
  border-radius: 4px;
  padding: 0.75rem 2.1rem 0.65rem;
  margin-top: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  min-width: 19.2rem;
}

.c7 {
  display: grid;
  place-items: center;
}

.c6 {
  border: none;
  background: none;
}

.c10 {
  border: none;
  background: none;
}

.c2 {
  border: 1px solid #e9eef2;
  border-radius: 4px;
  padding: 1.6rem;
  margin-top: 4px;
  margin-bottom: 1.6rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  font-family: 'Rubik';
  font-size: 2rem;
  line-height: 2.4rem;
  color: #4d6475;
  border: 0;
  outline: none;
}

.c3 {
  margin-right: 1.3rem;
}

@media (min-width:560px) {

}

@media (min-width:560px) {

}

@media (min-width:560px) {

}

@media (max-width:350px) {

}

@media (min-width:560px) {
  .c0 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    -ms-flex-pack: justify;
    justify-content: space-between;
  }
}

<div
  class="c0"
>
  <div>
    <h4
      class="c1"
      color="primaryBlack"
    >
      Total amount
    </h4>
    <div
      class="c2"
    >
      <img
        alt="Dollar Sign Icon"
        class="c3"
        src="[object Object]"
      />
      <input
        aria-label="Money input"
        class="c4"
        data-cy="money-input"
        inputmode="numeric"
        maxlength="28"
        placeholder="0.00"
        value="1,000.00"
      />
    </div>
  </div>
  <div>
    <h4
      class="c1"
      color="primaryBlack"
    >
      Reach goal by
    </h4>
    <div
      aria-label="Date picker"
      class="c5"
    >
      <button
        aria-label="Previous month"
        class="c6"
        data-cy="button-left"
      >
        <img
          alt="Arrow Left"
          src="[object Object]"
        />
      </button>
      <div
        aria-label="Month and year"
        class="c7"
        data-cy="date-values"
      >
        <h3
          class="c8"
          color="primaryBlack"
        >
          March
        </h3>
        <h3
          class="c9"
          color="primaryGray"
        >
          2021
        </h3>
      </div>
      <button
        aria-label="Next month"
        class="c10"
        data-cy="button-right"
      >
        <img
          alt="Arrow Right"
          src="[object Object]"
        />
      </button>
    </div>
  </div>
</div>
`
